File: LNCollectionView/LNScrollView/Simulator/LNScrollViewPageSimulator.m
//
//  LNScrollViewPageSimulator.m
//  LNCollectionView
//
//  Created by Levison on 12.11.24.
//

#import "LNScrollViewPageSimulator.h"

@interface LNScrollViewPageSimulator ()

@property (nonatomic, assign) CGFloat damping;
@property (nonatomic, assign) NSTimeInterval currentTime;
@property (nonatomic, assign) CGFloat velocity_0;
@property (nonatomic, assign) CGFloat offset;
@property (nonatomic, assign) CGFloat targetPosition;

@property (nonatomic, assign) CGFloat position;
@property (nonatomic, assign) CGFloat velocity;


@end

@implementation LNScrollViewPageSimulator

+ (CGFloat)targetOffsetWithVelocity:(CGFloat)velocity
                             offset:(CGFloat)offset
                            damping:(CGFloat)damping
{
    CGFloat currentTime = 1.f/(damping + velocity/offset);
    CGFloat v_0 = offset * exp(damping * currentTime)/currentTime;
    CGFloat maxPosition = v_0/(M_E * damping);
    return maxPosition;
}

- (instancetype)initWithPosition:(CGFloat)position
                        velocity:(CGFloat)velocity
                  targetPosition:(CGFloat)targetPosition
                         damping:(CGFloat)damping
{
    self = [super init];
    if (self) {
        self.damping = damping;
        self.targetPosition = targetPosition;
        self.offset = position - targetPosition;
        if (fabs(self.offset) < 1.f) {
            self.currentTime = 0.f;
            self.velocity_0 = velocity;
        } else {
            self.currentTime = 1.f/(self.damping + velocity/self.offset);
            self.velocity_0 = self.offset * exp(self.damping * self.currentTime)/self.currentTime;
        }
    }
    return self;
}

- (CGFloat)position {
    return self.targetPosition + self.offset;
}

- (void)accumulate:(NSTimeInterval)during {
    self.currentTime += during;
    self.offset = self.velocity_0 * self.currentTime * exp(- self.damping * self.currentTime);
    self.velocity = self.velocity_0 * exp(- self.damping * self.currentTime);
}

- (BOOL)isFinished
{
    return fabs(self.offset) < 0.1f && fabs(self.velocity) < 0.01;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/scrollview/simulator/LNScrollViewPageSimulator.kt
package com.ln.collectionview.scrollview.simulator

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.E

class LNScrollViewPageSimulator(
    initialPosition: Float,
    initialVelocity: Float,
    targetPosition: Float,
    damping: Float
) {

    var damping: Float = damping
    var currentTime: Float = 0f
    var velocity0: Float = 0f
    var offset: Float = initialPosition - targetPosition
    var targetPosition: Float = targetPosition
    var velocity: Float = initialVelocity
        private set

    init {
        if (abs(offset) < 1f) {
            currentTime = 0f
            velocity0 = initialVelocity
        } else {
            currentTime = 1f / (damping + initialVelocity / offset)
            velocity0 = offset * exp(damping * currentTime) / currentTime
        }
    }

    fun position(): Float {
        return targetPosition + offset
    }

    companion object {
        // 静态方法，计算目标偏移量
        fun targetOffsetWithVelocity(velocity: Float, offset: Float, damping: Float): Float {
            val currentTime = 1f / (damping + velocity / offset)
            val v0 = offset * exp(damping * currentTime) / currentTime
            return v0/(E.toFloat() * damping)
        }
    }

    fun accumulate(during: Float) {
        currentTime += during
        offset = velocity0 * currentTime * exp(-damping * currentTime)
        velocity = velocity0 * exp(-damping * currentTime)
    }

    fun isFinished(): Boolean {
        return abs(offset) < 0.1f && abs(velocity) < 0.01f
    }
}
